# Heart Disease Prediction
## Model Training and Evaluation

## 1. Import Libraries

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split

print("Libraries imported successfully!")

Libraries imported successfully!


## 2. Load Processed Dataset

In [10]:
import pandas as pd

url = "https://raw.githubusercontent.com/bashar730/heart-disease-ai/data-eda/data/processed/heart_disease_cleaned.csv"

df = pd.read_csv(url)

print("Processed dataset loaded successfully!")
df.head()

Processed dataset loaded successfully!


,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,63,1,1,145,233,1,2,150,0,2.3,3,0.0,6.0,0
1,67,1,4,160,286,0,2,108,1,1.5,2,3.0,3.0,1
2,67,1,4,120,229,0,2,129,1,2.6,2,2.0,7.0,1
3,37,1,3,130,250,0,0,187,0,3.5,3,0.0,3.0,0
4,41,0,2,130,204,0,2,172,0,1.4,1,0.0,3.0,0


In [ ]:
print("Dataset shape:", df.shape)

print("\nMissing values:")
print(df.isnull().sum())

print("\nTarget distribution:")
print(df["target"].value_counts())

Dataset shape: (303, 14)

Missing values:
age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

Target distribution:
target
0    164
1    139
Name: count, dtype: int64


## 3. Train/Test Split and Feature Scaling

In [13]:
from sklearn.model_selection import train_test_split
# Separate features and target
X = df.drop("target", axis=1)
y = df["target"]

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

X_train shape: (242, 13)
X_test shape: (61, 13)
y_train shape: (242,)
y_test shape: (61,)


In [ ]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

print("Feature scaling completed successfully!")
print("X_train_scaled shape:", X_train_scaled.shape)
print("X_test_scaled shape:", X_test_scaled.shape)

Feature scaling completed successfully!
X_train_scaled shape: (242, 13)
X_test_scaled shape: (61, 13)


## 4. Baseline Model Training and Evaluation

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Create and train the model
logistic_model = LogisticRegression(random_state=42)
logistic_model.fit(X_train_scaled, y_train)

# Make predictions
y_pred_logistic = logistic_model.predict(X_test_scaled)

# Evaluate the model
accuracy_logistic = accuracy_score(y_test, y_pred_logistic)

print("Logistic Regression Accuracy:", accuracy_logistic)
print("\nClassification Report:")
print(classification_report(y_test, y_pred_logistic))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_logistic))

Logistic Regression Accuracy: 0.8688524590163934

Classification Report:
              precision    recall  f1-score   support

           0       0.93      0.82      0.87        33
           1       0.81      0.93      0.87        28

    accuracy                           0.87        61
   macro avg       0.87      0.87      0.87        61
weighted avg       0.88      0.87      0.87        61


Confusion Matrix:
[[27  6]
 [ 2 26]]


In [ ]:
from sklearn.ensemble import RandomForestClassifier

# Create and train Random Forest model
rf_model = RandomForestClassifier(
    n_estimators=100,
    random_state=42
)

rf_model.fit(X_train, y_train)

# Make predictions
y_pred_rf = rf_model.predict(X_test)

# Evaluate model
accuracy_rf = accuracy_score(y_test, y_pred_rf)

print("Random Forest Accuracy:", accuracy_rf)

print("\nClassification Report:")
print(classification_report(y_test, y_pred_rf))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_rf))

Random Forest Accuracy: 0.8852459016393442

Classification Report:
              precision    recall  f1-score   support

           0       0.93      0.85      0.89        33
           1       0.84      0.93      0.88        28

    accuracy                           0.89        61
   macro avg       0.89      0.89      0.89        61
weighted avg       0.89      0.89      0.89        61


Confusion Matrix:
[[28  5]
 [ 2 26]]


In [ ]:
from sklearn.svm import SVC

# Create and train SVM model
svm_model = SVC(kernel='rbf', random_state=42)

svm_model.fit(X_train_scaled, y_train)

# Make predictions
y_pred_svm = svm_model.predict(X_test_scaled)

# Evaluate model
accuracy_svm = accuracy_score(y_test, y_pred_svm)

print("SVM Accuracy:", accuracy_svm)

print("\nClassification Report:")
print(classification_report(y_test, y_pred_svm))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_svm))

SVM Accuracy: 0.8524590163934426

Classification Report:
              precision    recall  f1-score   support

           0       0.90      0.82      0.86        33
           1       0.81      0.89      0.85        28

    accuracy                           0.85        61
   macro avg       0.85      0.86      0.85        61
weighted avg       0.86      0.85      0.85        61


Confusion Matrix:
[[27  6]
 [ 3 25]]


In [ ]:
from sklearn.neighbors import KNeighborsClassifier

# Create and train KNN model
knn_model = KNeighborsClassifier(n_neighbors=5)

knn_model.fit(X_train_scaled, y_train)

# Make predictions
y_pred_knn = knn_model.predict(X_test_scaled)

# Evaluate model
accuracy_knn = accuracy_score(y_test, y_pred_knn)

print("KNN Accuracy:", accuracy_knn)

print("\nClassification Report:")
print(classification_report(y_test, y_pred_knn))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_knn))

KNN Accuracy: 0.8852459016393442

Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.79      0.88        33
           1       0.80      1.00      0.89        28

    accuracy                           0.89        61
   macro avg       0.90      0.89      0.89        61
weighted avg       0.91      0.89      0.88        61


Confusion Matrix:
[[26  7]
 [ 0 28]]


In [ ]:
from sklearn.tree import DecisionTreeClassifier

# Create and train Decision Tree model
dt_model = DecisionTreeClassifier(random_state=42)

dt_model.fit(X_train, y_train)

# Make predictions
y_pred_dt = dt_model.predict(X_test)

# Evaluate model
accuracy_dt = accuracy_score(y_test, y_pred_dt)

print("Decision Tree Accuracy:", accuracy_dt)

print("\nClassification Report:")
print(classification_report(y_test, y_pred_dt))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_dt))

Decision Tree Accuracy: 0.7377049180327869

Classification Report:
              precision    recall  f1-score   support

           0       0.81      0.67      0.73        33
           1       0.68      0.82      0.74        28

    accuracy                           0.74        61
   macro avg       0.75      0.74      0.74        61
weighted avg       0.75      0.74      0.74        61


Confusion Matrix:
[[22 11]
 [ 5 23]]


In [ ]:
from sklearn.ensemble import GradientBoostingClassifier

# Create and train Gradient Boosting model
gb_model = GradientBoostingClassifier(random_state=42)

gb_model.fit(X_train, y_train)

# Make predictions
y_pred_gb = gb_model.predict(X_test)

# Evaluate model
accuracy_gb = accuracy_score(y_test, y_pred_gb)

print("Gradient Boosting Accuracy:", accuracy_gb)

print("\nClassification Report:")
print(classification_report(y_test, y_pred_gb))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_gb))

Gradient Boosting Accuracy: 0.8524590163934426

Classification Report:
              precision    recall  f1-score   support

           0       0.93      0.79      0.85        33
           1       0.79      0.93      0.85        28

    accuracy                           0.85        61
   macro avg       0.86      0.86      0.85        61
weighted avg       0.86      0.85      0.85        61


Confusion Matrix:
[[26  7]
 [ 2 26]]


## 5. Baseline Models Comparison

In [ ]:
from sklearn.metrics import precision_score, recall_score, f1_score

results = pd.DataFrame({
    "Model": [
        "Logistic Regression",
        "Random Forest",
        "SVM",
        "KNN",
        "Decision Tree",
        "Gradient Boosting"
    ],

    "Accuracy": [
        accuracy_logistic,
        accuracy_rf,
        accuracy_svm,
        accuracy_knn,
        accuracy_dt,
        accuracy_gb
    ],

    "Precision": [
        precision_score(y_test, y_pred_logistic),
        precision_score(y_test, y_pred_rf),
        precision_score(y_test, y_pred_svm),
        precision_score(y_test, y_pred_knn),
        precision_score(y_test, y_pred_dt),
        precision_score(y_test, y_pred_gb)
    ],

    "Recall": [
        recall_score(y_test, y_pred_logistic),
        recall_score(y_test, y_pred_rf),
        recall_score(y_test, y_pred_svm),
        recall_score(y_test, y_pred_knn),
        recall_score(y_test, y_pred_dt),
        recall_score(y_test, y_pred_gb)
    ],

    "F1-Score": [
        f1_score(y_test, y_pred_logistic),
        f1_score(y_test, y_pred_rf),
        f1_score(y_test, y_pred_svm),
        f1_score(y_test, y_pred_knn),
        f1_score(y_test, y_pred_dt),
        f1_score(y_test, y_pred_gb)
    ]
})

results = results.sort_values(
    by="Accuracy",
    ascending=False
).reset_index(drop=True)

results

,Model,Accuracy,Precision,Recall,F1-Score
0,Random Forest,0.885246,0.838710,0.928571,0.881356
1,KNN,0.885246,0.800000,1.000000,0.888889
2,Logistic Regression,0.868852,0.812500,0.928571,0.866667
3,SVM,0.852459,0.806452,0.892857,0.847458
4,Gradient Boosting,0.852459,0.787879,0.928571,0.852459
5,Decision Tree,0.737705,0.676471,0.821429,0.741935


## 6. Cross-Validation Evaluation

In [ ]:
from sklearn.model_selection import StratifiedKFold, cross_validate
from sklearn.pipeline import Pipeline
from sklearn.metrics import make_scorer, precision_score, recall_score, f1_score

cv = StratifiedKFold(
    n_splits=5,
    shuffle=True,
    random_state=42
)

cv_models = {
    "Logistic Regression": Pipeline([
        ("scaler", StandardScaler()),
        ("model", LogisticRegression(random_state=42))
    ]),

    "Random Forest": RandomForestClassifier(
        n_estimators=100,
        random_state=42
    ),

    "SVM": Pipeline([
        ("scaler", StandardScaler()),
        ("model", SVC(kernel="rbf", random_state=42))
    ]),

    "KNN": Pipeline([
        ("scaler", StandardScaler()),
        ("model", KNeighborsClassifier(n_neighbors=5))
    ]),

    "Decision Tree": DecisionTreeClassifier(
        random_state=42
    ),

    "Gradient Boosting": GradientBoostingClassifier(
        random_state=42
    )
}

scoring = {
    "accuracy": "accuracy",
    "precision": make_scorer(precision_score),
    "recall": make_scorer(recall_score),
    "f1": make_scorer(f1_score)
}

cv_results = []

for name, model in cv_models.items():

    scores = cross_validate(
        model,
        X,
        y,
        cv=cv,
        scoring=scoring
    )

    cv_results.append({
        "Model": name,
        "CV Accuracy": scores["test_accuracy"].mean(),
        "CV Precision": scores["test_precision"].mean(),
        "CV Recall": scores["test_recall"].mean(),
        "CV F1": scores["test_f1"].mean(),
        "Accuracy Std": scores["test_accuracy"].std()
    })

cv_results_df = pd.DataFrame(cv_results)

cv_results_df = cv_results_df.sort_values(
    by="CV F1",
    ascending=False
).reset_index(drop=True)

cv_results_df

,Model,CV Accuracy,CV Precision,CV Recall,CV F1,Accuracy Std
0,Random Forest,0.838142,0.854709,0.783862,0.816568,0.038786
1,SVM,0.838087,0.852606,0.783862,0.815278,0.039000
2,Logistic Regression,0.831585,0.830438,0.798413,0.812617,0.049559
3,KNN,0.821803,0.824888,0.777249,0.799504,0.026207
4,Gradient Boosting,0.795082,0.793492,0.754497,0.770804,0.053111
5,Decision Tree,0.765464,0.744874,0.762698,0.750771,0.046871


## 7. Hyperparameter Tuning

In [25]:
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestClassifier

param_grid_rf = {
    "n_estimators": [100, 200, 300],
    "max_depth": [None, 5, 10],
    "min_samples_split": [2, 5, 10],
    "min_samples_leaf": [1, 2, 4]
}

rf_grid = GridSearchCV(
    estimator=RandomForestClassifier(random_state=42),
    param_grid=param_grid_rf,
    cv=cv,
    scoring="f1",
    n_jobs=-1
)

rf_grid.fit(X_train, y_train)

print("Best Parameters:")
print(rf_grid.best_params_)

print("\nBest CV F1-Score:")
print(rf_grid.best_score_)

Best Parameters:
{'max_depth': 5, 'min_samples_leaf': 1, 'min_samples_split': 10, 'n_estimators': 100}

Best CV F1-Score:
0.7949049311530445


In [26]:
# Best tuned Random Forest model
best_rf = rf_grid.best_estimator_

# Predict on test data
y_pred_rf_tuned = best_rf.predict(X_test)

# Evaluate tuned model
accuracy_rf_tuned = accuracy_score(y_test, y_pred_rf_tuned)

print("Tuned Random Forest Accuracy:", accuracy_rf_tuned)

print("\nClassification Report:")
print(classification_report(y_test, y_pred_rf_tuned))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_rf_tuned))

Tuned Random Forest Accuracy: 0.9180327868852459

Classification Report:
              precision    recall  f1-score   support

           0       0.94      0.91      0.92        33
           1       0.90      0.93      0.91        28

    accuracy                           0.92        61
   macro avg       0.92      0.92      0.92        61
weighted avg       0.92      0.92      0.92        61


Confusion Matrix:
[[30  3]
 [ 2 26]]


In [18]:
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV
from sklearn.pipeline import Pipeline

svm_pipeline = Pipeline([
    ("scaler", StandardScaler()),
    ("svm", SVC())
])

param_grid_svm = {
    "svm__C": [0.1, 1, 10, 100],
    "svm__gamma": ["scale", "auto", 0.01, 0.1],
    "svm__kernel": ["rbf", "linear"]
}

svm_grid = GridSearchCV(
    estimator=svm_pipeline,
    param_grid=param_grid_svm,
    cv=cv,
    scoring="f1",
    n_jobs=-1
)

svm_grid.fit(X_train, y_train)

print("Best SVM Parameters:")
print(svm_grid.best_params_)

print("\nBest CV F1-Score:")
print(svm_grid.best_score_)

Best SVM Parameters:
{'svm__C': 1, 'svm__gamma': 0.01, 'svm__kernel': 'rbf'}

Best CV F1-Score:
0.8031223125512039


In [19]:
best_svm = svm_grid.best_estimator_

# Predict on test data
y_pred_svm_tuned = best_svm.predict(X_test)

# Evaluate tuned SVM
accuracy_svm_tuned = accuracy_score(y_test, y_pred_svm_tuned)

print("Tuned SVM Accuracy:", accuracy_svm_tuned)

print("\nClassification Report:")
print(classification_report(y_test, y_pred_svm_tuned))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_svm_tuned))

Tuned SVM Accuracy: 0.8524590163934426

Classification Report:
              precision    recall  f1-score   support

           0       0.88      0.85      0.86        33
           1       0.83      0.86      0.84        28

    accuracy                           0.85        61
   macro avg       0.85      0.85      0.85        61
weighted avg       0.85      0.85      0.85        61


Confusion Matrix:
[[28  5]
 [ 4 24]]


In [14]:
from sklearn.model_selection import GridSearchCV, StratifiedKFold
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
cv = StratifiedKFold(
    n_splits=5,
    shuffle=True,
    random_state=42
)

logistic_pipeline = Pipeline([
    ("scaler", StandardScaler()),
    ("logistic", LogisticRegression(
        random_state=42,
        max_iter=2000
    ))
])

param_grid_logistic = {
    "logistic__C": [0.01, 0.1, 1, 10, 100],
    "logistic__solver": ["liblinear", "lbfgs"],
    "logistic__class_weight": [None, "balanced"]
}

logistic_grid = GridSearchCV(
    estimator=logistic_pipeline,
    param_grid=param_grid_logistic,
    cv=cv,
    scoring="f1",
    n_jobs=-1
)

logistic_grid.fit(X_train, y_train)

print("Best Logistic Regression Parameters:")
print(logistic_grid.best_params_)

print("\nBest CV F1-Score:")
print(logistic_grid.best_score_)

Best Logistic Regression Parameters:
{'logistic__C': 10, 'logistic__class_weight': 'balanced', 'logistic__solver': 'liblinear'}

Best CV F1-Score:
0.8054167311229137


In [15]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
# Best tuned Logistic Regression model
best_logistic = logistic_grid.best_estimator_

# Predict on test data
y_pred_logistic_tuned = best_logistic.predict(X_test)

# Evaluate tuned model
accuracy_logistic_tuned = accuracy_score(
    y_test,
    y_pred_logistic_tuned
)

print("Tuned Logistic Regression Accuracy:", accuracy_logistic_tuned)

print("\nClassification Report:")
print(classification_report(y_test, y_pred_logistic_tuned))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_logistic_tuned))

Tuned Logistic Regression Accuracy: 0.8688524590163934

Classification Report:
              precision    recall  f1-score   support

           0       0.93      0.82      0.87        33
           1       0.81      0.93      0.87        28

    accuracy                           0.87        61
   macro avg       0.87      0.87      0.87        61
weighted avg       0.88      0.87      0.87        61


Confusion Matrix:
[[27  6]
 [ 2 26]]


In [ ]:
print("Final Logistic Regression Parameters:")
print("C =", loaded_model.named_steps["logistic"].C)
print("class_weight =", loaded_model.named_steps["logistic"].class_weight)
print("solver =", loaded_model.named_steps["logistic"].solver)
print("random_state =", loaded_model.named_steps["logistic"].random_state)
print("max_iter =", loaded_model.named_steps["logistic"].max_iter)

Final Logistic Regression Parameters:
C = 10
class_weight = balanced
solver = liblinear
random_state = 42
max_iter = 2000


In [21]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV
knn_pipeline = Pipeline([
    ("scaler", StandardScaler()),
    ("knn", KNeighborsClassifier())
])

param_grid_knn = {
    "knn__n_neighbors": [3, 5, 7, 9, 11, 13],
    "knn__weights": ["uniform", "distance"],
    "knn__p": [1, 2]
}

knn_grid = GridSearchCV(
    estimator=knn_pipeline,
    param_grid=param_grid_knn,
    cv=cv,
    scoring="f1",
    n_jobs=-1
)

knn_grid.fit(X_train, y_train)

print("Best KNN Parameters:")
print(knn_grid.best_params_)

print("\nBest CV F1-Score:")
print(knn_grid.best_score_)

Best KNN Parameters:
{'knn__n_neighbors': 3, 'knn__p': 1, 'knn__weights': 'uniform'}

Best CV F1-Score:
0.8155960412187708


In [22]:
# Best tuned KNN model
best_knn = knn_grid.best_estimator_

# Predict on test data
y_pred_knn_tuned = best_knn.predict(X_test)

# Evaluate tuned KNN
accuracy_knn_tuned = accuracy_score(
    y_test,
    y_pred_knn_tuned
)

print("Tuned KNN Accuracy:", accuracy_knn_tuned)

print("\nClassification Report:")
print(classification_report(y_test, y_pred_knn_tuned))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_knn_tuned))

Tuned KNN Accuracy: 0.8524590163934426

Classification Report:
              precision    recall  f1-score   support

           0       0.90      0.82      0.86        33
           1       0.81      0.89      0.85        28

    accuracy                           0.85        61
   macro avg       0.85      0.86      0.85        61
weighted avg       0.86      0.85      0.85        61


Confusion Matrix:
[[27  6]
 [ 3 25]]


In [27]:
from sklearn.model_selection import RepeatedStratifiedKFold, cross_validate

# Repeated Stratified Cross-Validation
repeated_cv = RepeatedStratifiedKFold(
    n_splits=5,
    n_repeats=10,
    random_state=42
)

final_models = {
    "Tuned Random Forest": best_rf,
    "Tuned SVM": best_svm,
    "Tuned Logistic Regression": best_logistic,
    "Tuned KNN": best_knn
}

scoring = {
    "accuracy": "accuracy",
    "precision": "precision",
    "recall": "recall",
    "f1": "f1"
}

final_cv_results = []

for name, model in final_models.items():

    scores = cross_validate(
        model,
        X,
        y,
        cv=repeated_cv,
        scoring=scoring,
        n_jobs=-1
    )

    final_cv_results.append({
        "Model": name,
        "Mean Accuracy": scores["test_accuracy"].mean(),
        "Accuracy Std": scores["test_accuracy"].std(),
        "Mean Precision": scores["test_precision"].mean(),
        "Mean Recall": scores["test_recall"].mean(),
        "Mean F1": scores["test_f1"].mean(),
        "F1 Std": scores["test_f1"].std()
    })

final_comparison = pd.DataFrame(final_cv_results)

final_comparison = final_comparison.sort_values(
    by="Mean F1",
    ascending=False
).reset_index(drop=True)

final_comparison

,Model,Mean Accuracy,Accuracy Std,Mean Precision,Mean Recall,Mean F1,F1 Std
0,Tuned SVM,0.842896,0.041961,0.869853,0.777037,0.818866,0.049080
1,Tuned Logistic Regression,0.831634,0.042713,0.825139,0.809339,0.814624,0.047696
2,Tuned KNN,0.823109,0.042466,0.810721,0.806640,0.806879,0.046221
3,Tuned Random Forest,0.830393,0.044939,0.851657,0.768280,0.805495,0.051351


In [28]:
from sklearn.model_selection import cross_validate

advanced_scoring = {
    "accuracy": "accuracy",
    "balanced_accuracy": "balanced_accuracy",
    "roc_auc": "roc_auc",
    "recall": "recall",
    "f1": "f1"
}

advanced_results = []

for name, model in final_models.items():

    scores = cross_validate(
        model,
        X,
        y,
        cv=repeated_cv,
        scoring=advanced_scoring,
        n_jobs=-1
    )

    advanced_results.append({
        "Model": name,
        "Mean Accuracy": scores["test_accuracy"].mean(),
        "Balanced Accuracy": scores["test_balanced_accuracy"].mean(),
        "ROC-AUC": scores["test_roc_auc"].mean(),
        "Recall": scores["test_recall"].mean(),
        "F1": scores["test_f1"].mean()
    })

advanced_comparison = pd.DataFrame(advanced_results)

advanced_comparison = advanced_comparison.sort_values(
    by="ROC-AUC",
    ascending=False
).reset_index(drop=True)

advanced_comparison

,Model,Mean Accuracy,Balanced Accuracy,ROC-AUC,Recall,F1
0,Tuned Random Forest,0.830393,0.825674,0.903280,0.768280,0.805495
1,Tuned SVM,0.842896,0.837998,0.903116,0.777037,0.818866
2,Tuned Logistic Regression,0.831634,0.830077,0.901882,0.809339,0.814624
3,Tuned KNN,0.823109,0.821956,0.870265,0.806640,0.806879


In [31]:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression

final_model = Pipeline([
    ("scaler", StandardScaler()),
    ("logistic", LogisticRegression(
        C=10,
        class_weight="balanced",
        solver="liblinear",
        random_state=42,
        max_iter=2000
    ))
])

# Train final model using the entire dataset
final_model.fit(X, y)

print("Final Logistic Regression model trained successfully!")
print("Number of training samples:", len(X))
print("Number of features:", X.shape[1])

Final Logistic Regression model trained successfully!
Number of training samples: 303
Number of features: 13


In [32]:
import joblib

# Save model together with the exact feature order
model_bundle = {
    "model": final_model,
    "feature_names": X.columns.tolist()
}

file_name = "heart_disease_model.pkl"

joblib.dump(model_bundle, file_name)

print("Model saved successfully!")
print("File name:", file_name)
print("\nFeature order:")
print(model_bundle["feature_names"])

Model saved successfully!
File name: heart_disease_model.pkl

Feature order:
['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal']


In [33]:
from google.colab import files

files.download("heart_disease_model.pkl")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [34]:
import joblib
import pandas as pd

# Load saved model bundle
bundle = joblib.load("heart_disease_model.pkl")

loaded_model = bundle["model"]
feature_names = bundle["feature_names"]

print("Model loaded successfully!")
print("Feature names:")
print(feature_names)

# Test with the first row from the dataset
sample = pd.DataFrame([{
    "age": 63,
    "sex": 1,
    "cp": 1,
    "trestbps": 145,
    "chol": 233,
    "fbs": 1,
    "restecg": 2,
    "thalach": 150,
    "exang": 0,
    "oldpeak": 2.3,
    "slope": 3,
    "ca": 0.0,
    "thal": 6.0
}])

prediction = loaded_model.predict(sample)[0]
probability = loaded_model.predict_proba(sample)[0][1]

print("\nTest prediction:", prediction)
print("Predicted probability:", probability)

Model loaded successfully!
Feature names:
['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal']

Test prediction: 0
Predicted probability: 0.3303672832815955


## Final Model Selection Summary

After training, evaluating, and tuning multiple machine learning models, Logistic Regression was selected as the final model.

The selection was based on its overall performance across the evaluation metrics, including Accuracy, Precision, Recall, F1-score, and ROC-AUC, as well as its simplicity and interpretability for heart disease risk prediction.

The final Logistic Regression model uses the following hyperparameters:

- C = 10
- class_weight = balanced
- solver = liblinear
- random_state = 42
- max_iter = 2000

The final model was trained on the complete processed dataset using all 13 input features and saved as:

`heart_disease_model.pkl`

The saved model was successfully reloaded and tested to verify that both prediction and probability estimation work correctly.

## Final Evaluation Results

The final Tuned Logistic Regression model achieved the following performance on the test set:

- **Accuracy:** 86.89%
- **Precision (Class 1):** 81%
- **Recall (Class 1):** 93%
- **F1-score (Class 1):** 87%
- **ROC-AUC (Test Set):** 95.02%
- **Test samples:** 61

### Confusion Matrix

| | Predicted 0 | Predicted 1 |
|---|---:|---:|
| Actual 0 | 27 | 6 |
| Actual 1 | 2 | 26 |

The model correctly identified 26 out of 28 positive heart disease cases, achieving a recall of 93% for Class 1.

This high recall is particularly important for heart disease risk prediction because reducing false negatives is important when identifying potentially at-risk cases.

The results, together with the model's interpretability and balanced overall performance, support the selection of Logistic Regression as the final model for integration into the Streamlit application.

In [ ]:
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score

# Load the processed dataset
url = "https://raw.githubusercontent.com/bashar730/heart-disease-ai/data-eda/data/processed/heart_disease_cleaned.csv"
df_roc = pd.read_csv(url)

# Separate features and target
X_roc = df_roc.drop("target", axis=1)
y_roc = df_roc["target"]

# Same train/test split used in the project
X_train_roc, X_test_roc, y_train_roc, y_test_roc = train_test_split(
    X_roc,
    y_roc,
    test_size=0.2,
    random_state=42,
    stratify=y_roc
)

# Rebuild the selected tuned Logistic Regression model
roc_model = Pipeline([
    ("scaler", StandardScaler()),
    ("logistic", LogisticRegression(
        C=10,
        class_weight="balanced",
        solver="liblinear",
        random_state=42,
        max_iter=2000
    ))
])

# Train only on the training set
roc_model.fit(X_train_roc, y_train_roc)

# Probability predictions on the untouched test set
y_prob_roc = roc_model.predict_proba(X_test_roc)[:, 1]

# Calculate ROC-AUC
roc_auc_final = roc_auc_score(y_test_roc, y_prob_roc)

print("Final Logistic Regression Test ROC-AUC:", roc_auc_final)

Final Logistic Regression Test ROC-AUC: 0.9502164502164502


In [ ]:
import joblib

# Load the final saved model
bundle = joblib.load("heart_disease_model.pkl")
final_model = bundle["model"]

print("Final saved model:")
print(final_model)

print("\nFinal Logistic Regression Parameters:")
print("C =", final_model.named_steps["logistic"].C)
print("class_weight =", final_model.named_steps["logistic"].class_weight)
print("solver =", final_model.named_steps["logistic"].solver)
print("random_state =", final_model.named_steps["logistic"].random_state)
print("max_iter =", final_model.named_steps["logistic"].max_iter)

print("\nNumber of input features:", len(bundle["feature_names"]))
print("Feature names:", bundle["feature_names"])

Final saved model:
Pipeline(steps=[('scaler', StandardScaler()),
                ('logistic',
                 LogisticRegression(C=10, class_weight='balanced',
                                    max_iter=2000, random_state=42,
                                    solver='liblinear'))])

Final Logistic Regression Parameters:
C = 10
class_weight = balanced
solver = liblinear
random_state = 42
max_iter = 2000

Number of input features: 13
Feature names: ['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal']


In [ ]:
from google.colab import drive
import os
import shutil

drive.mount('/content/drive')

save_dir = "/content/drive/MyDrive/Heart_Disease_Project/models"
os.makedirs(save_dir, exist_ok=True)

source = "/content/heart_disease_model.pkl"
destination = os.path.join(save_dir, "heart_disease_model.pkl")

shutil.copy2(source, destination)

print("Model saved successfully to:")
print(destination)

MessageError: Error: credential propagation was unsuccessful